File: python/adapter/adapter.py
from abc import ABC


class Duck(ABC):
    def fly(self):
        print("Flying!")

    def quack(self):
        print("Quacking!")


class ITurkey(ABC):
    def fly(self):
        pass

    def gobble(self):
        pass


class Turkey(Duck):
    def fly(self):
        super().fly()

    def quack(self):
        print("Gobbling!")


class TurkeyAdapter(ITurkey):
    def __init__(self, duck: Duck):
        self._duck = duck

    def fly(self):
        self._duck.fly()

    def gobble(self):
        self._duck.quack()


def main():
    turkey = Turkey()
    adapter = TurkeyAdapter(turkey)
    adapter.fly()
    adapter.gobble()


if __name__ == '__main__':
    main()

File: cpp/adapter/adapter.cpp
#include <iostream>
#include <memory>
#include <utility>

class Duck {
public:
    virtual void fly() {
        std::cout << "Flying!" << std::endl;
    }

    virtual void quack() {
        std::cout << "Quacking!" << std::endl;
    }

protected:
    Duck() = default;
};

class ITurkey {
public:
    virtual void fly() = 0;

    virtual void gobble() = 0;
};

class Turkey : public Duck {
    void quack() override {
        std::cout << "Gobbling!" << std::endl;
    }
};

class TurkeyAdapter : public ITurkey {
public:
    explicit TurkeyAdapter(std::shared_ptr<Duck> duck) {
        _duck = std::move(duck);
    }

    void fly() override {
        _duck->fly();
    }

    void gobble() override {
        _duck->quack();
    }

private:
    std::shared_ptr<Duck> _duck;
};

int main() {
    TurkeyAdapter adapter(std::make_shared<Turkey>());
    adapter.fly();
    adapter.gobble();
    return 0;
}